### ISSUE TITLE
Unhelpful error when trying to import from a crate without 'lib' crate-type

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust#96039]
When working on a crate using `crate-type = ["cdylib"]` or `crate-type = ["staticlib"]`¹, one may want to also write unit tests (or examples) for the public functions in that crate. This "simply" requires adding the default `"lib"` crate-type back into the `Cargo.toml`.

However, discovering that isn't exactly straight-forward. Cargo doesn't complain when there's tests or examples in one such crate but once you try to import something from the crate you want to test, you get:

```
error[E0432]: unresolved import `the_crate`
 --> tests/foo.rs:1:5
  |
1 | use the_crate::SomeType;
  |     ^^^^^^^^^ use of undeclared crate or module `the_crate`
  |
```

Given Rust's excellent diagnostics elsewhere, I would expect either

* The presence of a test or example on a non-`lib` crate to raise a warning in addition to this error (this would probably be a simple Cargo change, but might also be annoying to some projects)
* The error to include a hint about `the_crate` existing, but not being available due to its `crate-type` (this might require cargo passing more info to rustc first?)

¹ plugins for applications, or libraries meant to be used in codebases written in a different language than Rust

**Comment by WilliamForsdal** [comment]
Just ran into this as well. I'm building a cdylib and I noticed that the final dll size is increased by about 300kB when adding "lib", so it's a tradeoff.

**Comment by pmj** [comment]
Thanks @jplatte for filing this, I've just spent half an hour running around in circles chasing a "use of undeclared crate or module" error and it would have taken me longer to fix had I not found this report.

I set up a library crate and a binary crate. I've been prototyping in the binary, the library is intended for use as part of a (non-Rust) macOS application and thus was set up as a `crate-type = ["staticlib"]`. Then I started moving code from the binary crate into the library and wanted to link the library into the binary as well, and I got that error despite specifying the dependency, etc.

Changing the library crate to `crate-type = ["staticlib", "lib"]` fixed it.

Some kind of "wrong crate type" error would be much more helpful, but I'm thinking it might actually be cargo that's the problem here because `cargo build --verbose` reveals that there's no `--extern <LIBCRATENAME>=<path/to/.rlib>` argument passed to `rustc` for that dependency. So really cargo ought to detect the problem when attempting to resolve the dependency, but I guess it might be tricky because maybe there is a reason you might genuinely want to define a non-`lib` crate as a dependency?

**Comment by Enselic** [comment]
Triage: As far as I can tell this is a cargo issue related to how it builds things and what args it passes to rustc, because rustc itself is capable of giving a good error message. This:
```sh
cd $(mktemp -d) # Go to a new empty dir
echo 'pub fn foo() {}' > foo.rs
rustc --crate-type staticlib foo.rs
echo 'extern crate foo; fn main() { foo::foo(); }' > main.rs
rustc -L. main.rs
```
gives:
```
error[E0462]: found staticlib `foo` instead of rlib or dylib
 --> main.rs:1:1
  |
1 | extern crate foo; fn main() { foo::foo(); }
  | ^^^^^^^^^^^^^^^^^
  |
  = note: the following crate versions were found:
          crate `foo`: ./libfoo.a
  = help: please recompile that crate using --crate-type lib

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0462`.
```

which explicitly mentions that crate-type `lib` is missing.

Moving to cargo.

@rustbot transfer cargo
